pageserver: refactor ingest inplace to decouple decoding and handling

VladLazar opened the pull request:
## Problem

WAL ingest couples decoding of special records with their handling (updates to the storage engine mostly).
This is a roadblock for our plan to move WAL filtering (and implicitly decoding) to safekeepers since they cannot
do writes to the storage engine. 

## Summary of changes

This PR decouples the decoding of the special WAL records from their application. The changes are done in place
and I've done my best to refrain from refactorings and attempted to preserve the original code as much as possible.
The commit messages flag deviations.

## Reviewer Note

You will notice that this is a somewhat transient state. This is intentional. I'd like this review to focus on ensuring
that I've not changed the handling of these special records.

Future PRs will:
* Move all the new types and decoding logic into a crate of its own
* Create separate instances for these types for each supported pg version
* Further refactor ingest to take a `SerializedBatch` and a series of special records for handling (types defined in this PR)

Related: https://github.com/neondatabase/neon/issues/9335
Epic: https://github.com/neondatabase/neon/issues/9329

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

problame reviewed (commented):
No good tooling to post review comments for individual commits (I reviewed commit by commit).
So, dumping comments here

---

```
         return Ok(Some(SmgrRecord::Truncate(SmgrTruncate {
                    rel,
                    to: truncate.blkno,
                })));
            }

            if (truncate.flags & pg_constants::SMGR_TRUNCATE_FSM) != 0 {
                let rel = RelTag {
                    spcnode,
```

In the original code, theoretically `SMGR_TRUNCATE_HEAP`  and `SMGR_TRUNCATE_FSM` could both be set, so, it's not technically equivalent to return on the first matching bit. I guess in practice they're exclusive though? => should `anyhow::ensure!` that at the top of the function.

---

```
   let (xact_common, is_commit, is_prepared) = match record {
            XactRecord::Prepare(XactPrepare { xl_xid, data }) => {
                let xid: u64 = if modification.tline.pg_version >= 17 {
                    self.adjust_to_full_transaction_id(xl_xid)?
                } else {
                    xl_xid as u64
                };
                return modification.put_twophase_file(xid, data, ctx).await;
            }
```

Before this PR, this special case was for the twophase PREPARED records, i.e., `XLOG_XACT_COMMIT_PREPARED` and `XLOG_XACT_ABORT_PREPARED`. This PR moves that handling to the `XactRecord::Prepare`, which afaict is the wrong place. Needs to be in the cases for `XactRecord::CommitPrepared` and `XactRecord::AbortPrepared`.

---

Please create a follow-up task to address the `.unwrap().unwrap()` business that multiple commits added, and link it from the epic & this PR.
If we don't address this, inevitably eventually we'll make stuff fallible and cause a panic.

---

```

    fn decode_logical_message_record(
        buf: &mut Bytes,
        decoded: &DecodedWALRecord,
        _pg_version: u32,
    ) -> anyhow::Result<Option<LogicalMessageRecord>> {
        let info = decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK;
        if info == pg_constants::XLOG_LOGICAL_MESSAGE {
            let xlrec = crate::walrecord::XlLogicalMessage::decode(buf);
            let buf_size = xlrec.prefix_size + xlrec.message_size;
            // TODO change decode function interface to take ownership of buf
```

I think it would be more natural, robust, and efficient, to filter the `prefix` during decode instead of during ingest, which is what this PR does right now.
It would also allow us to have an enum for the two variants (`neon-test` and `neon-file:`).

Independent of that discussion, did you check the curren tusers of `neon-test` wrt where the failpoint happens?
Like, probably all the tests that use it just expect that they can hold up pageserver walreceiver via the `wal-ingest-logical-message-sleep` .
But would make sense to check that / maybe rename the failpoint to `pageserver-wal-ingest-...` to make things a bit clearer.

---

Have you thought through what happens if the decoding sides pg_version (=SK's view of the world) gets out of sync with the ingesting side's pg_version  (=PS's view of the world?
Do we want to protect against that by including the pg_version in each of the decoded records?
=> ok to do this later

---

```
 async fn ingest_heapam_record(
        &mut self,
        clear_vm_bits: ClearVmBits,
        modification: &mut DatadirModification<'_>,
        ctx: &RequestContext,
    ) -> anyhow::Result<()> {
        let ClearVmBits {
```

IMO this should still take a `HeapamRecord` to get more type safety.
Even though that enum only has one variant, `ClearVmBits`.
You can still destructure it painfree.


Same for `ingest_neonrmgr_record`

---


Hmm, the last commit that DRYs the clear VM bits code should be a separate PR IMO.
Can't speak to whether it's a wise choice or not to DRY the code.
Matthias would know, I think.

VladLazar commented:
> ```
>          return Ok(Some(SmgrRecord::Truncate(SmgrTruncate {
>                     rel,
>                     to: truncate.blkno,
>                 })));
>             }
> 
>             if (truncate.flags & pg_constants::SMGR_TRUNCATE_FSM) != 0 {
>                 let rel = RelTag {
>                     spcnode,
> ```
> In the original code, theoretically `SMGR_TRUNCATE_HEAP` and `SMGR_TRUNCATE_FSM` could both be set, so, it's not technically equivalent to return on the first matching bit. I guess in practice they're exclusive though? => should `anyhow::ensure!` that at the top of the function.

Thank you for spotting this. I've adjusted the handling to allow for truncation of multiple fork types in the same record in https://github.com/neondatabase/neon/pull/9472/commits/4580b545a15cf4fa913886153b07787d41a12363.

> ```
>    let (xact_common, is_commit, is_prepared) = match record {
>             XactRecord::Prepare(XactPrepare { xl_xid, data }) => {
>                 let xid: u64 = if modification.tline.pg_version >= 17 {
>                     self.adjust_to_full_transaction_id(xl_xid)?
>                 } else {
>                     xl_xid as u64
>                 };
>                 return modification.put_twophase_file(xid, data, ctx).await;
>             }
> ```
> 
> Before this PR, this special case was for the twophase PREPARED records, i.e., `XLOG_XACT_COMMIT_PREPARED` and `XLOG_XACT_ABORT_PREPARED`. This PR moves that handling to the `XactRecord::Prepare`, which afaict is the wrong place. Needs to be in the cases for `XactRecord::CommitPrepared` and `XactRecord::AbortPrepared`.

I don't see it, but perhaps I'm missing something here. Let's work through it.

1. XLOG_XACT_PREPARE
  a. Previous code: It was handled [here](https://github.com/neondatabase/neon/blob/0595320c87c4bcab9e346cf7904f7e4b00454388/pageserver/src/walingest.rs#L359C1-L368) by: adjusting the transaction id and `calling put_twophase_file`.
  b. New code: It is now handled [here](https://github.com/neondatabase/neon/blob/991a4c00b84e181cac0acd59e197ac20611340c2/pageserver/src/walingest.rs#L1558-L1565) with exactly the same code as previously
2. XLOG_XACT_ABORT_PREPARED or XLOG_XACT_COMMIT_PREPARED
  a. Previous code: It was handled [here](https://github.com/neondatabase/neon/blob/0595320c87c4bcab9e346cf7904f7e4b00454388/pageserver/src/walingest.rs#L332-L358) by: ingesting the xact record, adjusting xid and, finally, dropping twophase files
  b. New code: It is now handled [here](https://github.com/neondatabase/neon/blob/991a4c00b84e181cac0acd59e197ac20611340c2/pageserver/src/walingest.rs#L1572-L1667) with `is_prepared=true`. It's hard to tell from the diff but the code block linked
  previously merges two pre-existint code blocks: [previous body of ingest_xact_record](https://github.com/neondatabase/neon/blob/0595320c87c4bcab9e346cf7904f7e4b00454388/pageserver/src/walingest.rs#L1494-L1573) and [epilogue of the old handling which does the xid adjustment and dropping of two phase files](https://github.com/neondatabase/neon/blob/0595320c87c4bcab9e346cf7904f7e4b00454388/pageserver/src/walingest.rs#L346-L358). In the new code, we do xid adjustment and dropping of two phase files at the [end of ingest_xact_record](https://github.com/neondatabase/neon/blob/991a4c00b84e181cac0acd59e197ac20611340c2/pageserver/src/walingest.rs#L1650-L1665) for XLOG_XACT_ABORT_PREPARED or XLOG_XACT_COMMIT_PREPARED.

To me, this looks equiavalent to the old code, but let me know if you're not convinced.

> Please create a follow-up task to address the `.unwrap().unwrap()` business that multiple commits added, and link it from the epic & this PR. If we don't address this, inevitably eventually we'll make stuff fallible and cause a panic.

I already have a [draft of a follow-follow-up PR](https://github.com/neondatabase/neon/pull/9487) where this is fixed, but sure: https://github.com/neondatabase/neon/issues/9491.

> ```
> 
>     fn decode_logical_message_record(
>         buf: &mut Bytes,
>         decoded: &DecodedWALRecord,
>         _pg_version: u32,
>     ) -> anyhow::Result<Option<LogicalMessageRecord>> {
>         let info = decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK;
>         if info == pg_constants::XLOG_LOGICAL_MESSAGE {
>             let xlrec = crate::walrecord::XlLogicalMessage::decode(buf);
>             let buf_size = xlrec.prefix_size + xlrec.message_size;
>             // TODO change decode function interface to take ownership of buf
> ```
> 
> I think it would be more natural, robust, and efficient, to filter the `prefix` during decode instead of during ingest, which is what this PR does right now. It would also allow us to have an enum for the two variants (`neon-test` and `neon-file:`).

I like the suggestion. Done in https://github.com/neondatabase/neon/pull/9472/commits/4f8c733ec0ac0bb6880a443698120bcf136adfa7.

> Independent of that discussion, did you check the curren tusers of `neon-test` wrt where the failpoint happens? Like, probably all the tests that use it just expect that they can hold up pageserver walreceiver via the `wal-ingest-logical-message-sleep` . But would make sense to check that / maybe rename the failpoint to `pageserver-wal-ingest-...` to make things a bit clearer.

Renamed the failpoint in https://github.com/neondatabase/neon/pull/9472/commits/9c7baecdd294c2ecf0567d3b388a848c84827108.

> Have you thought through what happens if the decoding sides pg_version (=SK's view of the world) gets out of sync with the ingesting side's pg_version (=PS's view of the world? Do we want to protect against that by including the pg_version in each of the decoded records? => ok to do this later

Subtle breakage. Will work fine for 90% of records until you get to one with PG specific parsing or handling.
If we detect that (SK can include it's own pg version for the timeline), we should kill the wal connection and drop our current in-memory file (like any other error on the ingest path).
This is not really consistent right now, so will have to be thought through more and tested.

> ```
>  async fn ingest_heapam_record(
>         &mut self,
>         clear_vm_bits: ClearVmBits,
>         modification: &mut DatadirModification<'_>,
>         ctx: &RequestContext,
>     ) -> anyhow::Result<()> {
>         let ClearVmBits {
> ```
> 
> IMO this should still take a `HeapamRecord` to get more type safety. Even though that enum only has one variant, `ClearVmBits`. You can still destructure it painfree.
> 
> Same for `ingest_neonrmgr_record`

I don't see how that's more type safe. Are you alluding to the scenario where we call `ingest_heapam_record` from the neonrmgr handling? That's why I want to keep the DRI
change for `ClearVmBits`.

> Hmm, the last commit that DRYs the clear VM bits code should be a separate PR IMO. Can't speak to whether it's a wise choice or not to DRY the code. Matthias would know, I think.

See above.

problame reviewed (approved):
> To me, this looks equiavalent to the old code, but let me know if you're not convinced.

Ok, I hadn't seen the inlining / DRYing of the body of ingest_xact_record.

hlinnaka commented on pageserver/src/walingest.rs:
```diff
@@ -107,6 +107,139 @@ struct WarnIngestLag {
     timestamp_invalid_msg_ratelimit: RateLimit,
 }
 
+enum HeapamRecord {
```
A high level comment that explains what these structs are for would be nice. Something like:

```
// These structs are an intermediary representation of the PostgreSQL WAL records.
```

hlinnaka replied:
These structs seem similar to the `XlRelmapUpdate`, `XlHeapDelete` etc. structs in `walrecord.rs`. Are these exactly the same idea? Maybe put these in walrecord.rs, too?

VladLazar replied:
> These structs seem similar to the `XlRelmapUpdate`, `XlHeapDelete` etc. structs in walrecord.rs.

They are similar indeed. Think of these new structs as higher level constructs: they have all the context required for the PS to act on them. In some cases that was already true for the `Xl...` types as well and in some cases it was not.

The next PR in the series cleans this up a bit by moving all the `Xl...` types into `postgres_ffi` and all of these "higher level representations" in a new `wal_decoder` module.

VladLazar replied:
Heikki, we had a brief out of band chat about this work, but I don't think I mentioned the [RFC](https://github.com/neondatabase/neon/pull/8754). Might be good for some extra context.

VladLazar replied:
Added a comment in https://github.com/neondatabase/neon/pull/9472/commits/02fbac3f58495ec650909155743c4b17a87af033
